python/mlc_chat/compiler/compiler_pass/library_dispatch.py

```diff
@@ -0,0 +1,121 @@
+
+        Q_BNSH = is_op("relax.permute_dims")(Q)
+
+        if use_flash_mqa:
+            K_BNSH = is_op("relax.permute_dims")(is_op("relax.repeat")(K))
+            V_BNSH = is_op("relax.permute_dims")(is_op("relax.repeat")(V))
+        else:
+            K_BNSH = is_op("relax.permute_dims")(K)
+            V_BNSH = is_op("relax.permute_dims")(V)
+
+        K_BNSH_T = is_op("relax.permute_dims")(K_BNSH)
+
+        matmul1 = is_op("relax.matmul")(Q_BNSH, K_BNSH_T)
+        divide = is_op("relax.divide")(matmul1, is_const())
+        max = is_op("relax.maximum")(divide, is_const())
+        min = is_op("relax.minimum")(max, wildcard())
+        softmax = is_op("relax.nn.softmax")(is_op("relax.astype")(min))
+        matmul2 = is_op("relax.matmul")(is_op("relax.astype")(softmax), V_BNSH)
+
+        pattern = is_op("relax.permute_dims")(matmul2)
+
+        def callback(_, matchings):
+            return R.nn.attention(
+                matchings[Q], matchings[K], matchings[V], causal_mask="BottomRight"
+            )
+
+        new_module = {}
+        for gvar, func in mod.functions.items():
+            if isinstance(func, tvm.relax.Function):
+                func = rewrite_call(pattern, callback, func)
+            new_module[gvar] = func
+
+        return tvm.IRModule(new_module, mod.type_definitions, mod.attrs, mod.global_infos)
+
+    return ir_module_transform
+
+
+@tvm.transform.module_pass(opt_level=0, name="LibraryDispatch")
+class LibraryDispatch:  # pylint: disable=too-few-public-methods
+    """A compiler pass that fuses transpose + dequantize."""
+
+    def __init__(self, args: CompileArgs) -> None:
+        self.args = args
+
+    def transform_module(self, mod: IRModule, _ctx: tvm.transform.PassContext) -> IRModule:
+        """IRModule-level transformation"""
+        if self.args.target.kind.name != "cuda":
+            return mod
+
+        patterns = []
+
+        has_cutlass = tvm.get_global_func("relax.ext.cutlass", True)
+
+        if has_cutlass and self.args.opt.cutlass_attn:
+            # pylint: disable=no-value-for-parameter
+            mod = rewrite_attention(use_flash_mqa=False)(mod)
+            patterns += get_patterns_with_prefix("cutlass.attention")
+
+        if has_cutlass and self.args.opt.cutlass_norm:
+            patterns += get_patterns_with_prefix("cutlass.layer_norm")
+            patterns += get_patterns_with_prefix("cutlass.rms_norm")
+
+        has_cublas = tvm.get_global_func("relax.ext.cublas", True)
+
+        if (
+            has_cublas
+            and self.args.quantization.name in ("q0f16", "q0f32")
+            and not self.args.opt.cublas_gemm
+        ):
+            patterns += get_patterns_with_prefix("cublas")
+
+        if len(patterns) > 0:
+            os.makedirs("./tmp", exist_ok=True)
+
+            major, minor = parse_compute_version(tvm.cuda(0).compute_version)
+
+            if major == 8:
+                sm = 80
+            else:
+                sm = 10 * major + minor
```

Comment:
We have to deal with cases where sm is not 7x or 8x (e.g. 6x or H100 which is sm_90)